storage_service/locations/models/duracloud.py

```diff
@@ -158,36 +253,93 @@ def move_to_storage_service(self, src_path, dest_path, dest_space):
-                LOGGER.debug('Writing %s to %s', entry, dest)
-                with open(dest, 'wb') as f:
-                    f.write(response.content)
-        elif response.status_code != 200:
-            raise StorageException('Unable to fetch %s' % src_path)
-        else:  # status_code == 200
-            self.space._create_local_directory(dest_path)
-            LOGGER.debug('Writing to %s', dest_path)
-            with open(dest_path, 'wb') as f:
-                f.write(response.content)
+                dest = entry.replace(src_path, dest_path, 1)
+                self._download_file(url, dest)
 
     def _upload_file(self, url, upload_file):
-        # Example URL: https://trial.duracloud.org/durastore/trial261//ts/test.txt
-        with open(upload_file, 'rb') as f:
-            response = self.session.put(url, data=f)
-        LOGGER.info('Response from %s: %s', url, response)
-        if response.status_code != 201:
-            LOGGER.warning('Response text: %s', response.text)
-            raise StorageException('Unable to store %s' % upload_file)
+        """
+        Helper to upload files to Duracloud.
+
+        :param url: URL to upload the file to.
+        :param upload_file: Absolute path to the file to upload.
+        :returns: None
+        :raises: StorageException if error storing file
+        """
+        LOGGER.debug('Upload %s to %s', upload_file, url)
+        filesize = os.path.getsize(upload_file)
+        if filesize > self.CHUNK_SIZE:
+            LOGGER.debug('%s size (%s) larger than %s', upload_file, filesize, self.CHUNK_SIZE)
+            # Create manifest info for complete file.  Eg:
+            # <header schemaVersion="0.2">
+            #   <sourceContent contentId="chunked/chunked_image.jpg">
+            #     <mimetype>application/octet-stream</mimetype>
+            #     <byteSize>2222135</byteSize>
+            #     <md5>9497f70a1a17943ddfcbed567538900d</md5>
+            #   </sourceContent>
+            # </header>
+            relative_path = urllib.unquote(url.replace(self.duraspace_url, '', 1))
+            LOGGER.debug('File name: %s', relative_path)
+            checksum = utils.generate_checksum(upload_file, 'md5')
+            root = etree.Element('{duracloud.org}chunksManifest', nsmap={'dur': 'duracloud.org'})
+            header = etree.SubElement(root, 'header', schemaVersion="0.2")
+            content = etree.SubElement(header, 'sourceContent', contentId=relative_path)
+            etree.SubElement(content, 'mimetype').text = 'application/octet-stream'
+            etree.SubElement(content, 'byteSize').text = str(filesize)
+            etree.SubElement(content, 'md5').text = checksum.hexdigest()
+            chunks = etree.SubElement(root, 'chunks')
+            # Split file into chunks
+            with open(upload_file, 'rb') as f:
+                i = 0
+                chunk_data = f.read(self.CHUNK_SIZE)
+                while chunk_data:
```

Comment:
Nitpick, but do we need to use the filesystem as staging for these chunks? If we're already reading them into memory, can we just checksum and then upload them from memory instead?